homeassistant/components/volvooncall.py

```diff
@@ -85,52 +70,76 @@ def setup(hass, config):
 
     def discover_vehicle(vehicle):
         """Load relevant platforms."""
-        state.entities[vehicle.vin] = []
-        for attr, (component, *_) in RESOURCES.items():
-            if (getattr(vehicle, attr + '_supported', True) and
-                    attr in config[DOMAIN].get(CONF_RESOURCES, [attr])):
-                discovery.load_platform(
-                    hass, component, DOMAIN, (vehicle.vin, attr), config)
+        state.vehicles.append(vehicle.vin)
 
-    def update_vehicle(vehicle):
-        """Receive updated information on vehicle."""
-        state.vehicles[vehicle.vin] = vehicle
-        if vehicle.vin not in state.entities:
-            discover_vehicle(vehicle)
+        dashboard = vehicle.dashboard(
+            mutable=config[DOMAIN].get(CONF_MUTABLE),
+            scandinavian_miles=config[DOMAIN].get(CONF_SCANDINAVIAN_MILES))
 
-        for entity in state.entities[vehicle.vin]:
-            entity.schedule_update_ha_state()
+        def is_enabled(attr):
+            return attr in config[DOMAIN].get(CONF_RESOURCES, [attr])
 
-        dispatcher_send(hass, SIGNAL_VEHICLE_SEEN, vehicle)
+        for instrument in (
+                instrument
+                for instrument in dashboard.instruments
+                if is_enabled(instrument.slug_attr)):
 
-    def update(now):
+            state.instruments.append(instrument)
+
+            hass.async_create_task(
+                discovery.async_load_platform(
+                    hass,
+                    instrument.component,
+                    DOMAIN,
+                    (vehicle.vin,
+                     instrument.component,
+                     instrument.attr),
+                    config))
+
+    async def update(now):
         """Update status from the online service."""
         try:
-            if not connection.update():
+            if not await connection.update(journal=True):
                 _LOGGER.warning("Could not query server")
                 return False
 
             for vehicle in connection.vehicles:
-                update_vehicle(vehicle)
+                if vehicle.vin not in state.vehicles:
+                    discover_vehicle(vehicle)
+
+            for entity in state.entities:
+                entity.async_schedule_update_ha_state()
```

Comment:
Entities are only allowed to be accessed from within their respective platforms. Use the dispatch helper to signal the correct entity from the component instead.

Reply:
Ok, you mean `async_dispatcher_send(hass, SIGNAL_STATE_UPDATED)`  here and then in the entity: `async_dispatcher_connect(hass, SIGNAL_STATE_UPDATED, self.async_schedule_update_ha_state)`?